src/main/java/com/louis/BinarySearchTree.java

```diff
@@ -0,0 +1,283 @@
+// CHECKSTYLE:OFF
+
+// CHECKSTYLE:ON
+
+package com.louis;
+
+import com.google.common.base.Preconditions;
+import java.util.NoSuchElementException;
+
+/**
+ * A binary search tree implementation.
+ * <p>
+ * Binary search trees (BSTs) are derived from the associative array abstract data type. It is a symbol table with keys
+ * and values; null is excluded. The tree is not balanced. Ks are Comparable and the tree is sorted by them. Most
+ * operations have linear worst case complexity.
+ *
+ * @author Louis Ashton (louisashton@live.com)
+ */
+public class BinarySearchTree<K extends Comparable<K>, V> {
+
+    // The root of the binary search tree.
+    private Node root;
+
+    /**
+     * A node of a binary search tree.
+     * <p>
+     * Each node has a key and associated value. Pointers to the node's children are also stored.
+     */
+    private class Node {
+        private K key;
+        private V value;
+        private Node leftSubtree;
+        private Node rightSubtree;
+        private int sizeOfSubtree;
+
+        /**
+         * Creates a Node.
+         */
+        Node(K key, V value, int sizeOfSubtree) {
+            this.key = key;
+            this.value = value;
+            this.sizeOfSubtree = sizeOfSubtree;
+        }
+    }
+
+    /**
+     * Creates a binary search tree.
+     */
+    public BinarySearchTree() {}
+
+    /**
+     * Returns the number of nodes in the binary search tree.
+     */
+    public final int size() {
+        return size(root);
+    }
+
+    /**
+     * Gets the size of a subtree.
+     *
+     * @param node the node is the root of the subtree.
+     * @return Returns the number of pairs descendant from a node
+     */
+    private int size(Node node) {
```

Comment:
consider `calculateSizeOfSubtreeFromNode(Node node)` as the method name for clarity.